tests/components/room/action.service.spec.js

```diff
@@ -0,0 +1,665 @@
+      sendChatMessage: jasmine.createSpy('DataChannelService.sendChatMessage')
+    };
+  });
+
+  // Mock External modules
+  beforeEach(function () {
+    module(function($provide) {
+      $provide.value('Feed', FeedSpy);
+      $provide.value('FeedsService', FeedsServiceSpy);
+      $provide.value('LogEntry', LogEntrySpy);
+      $provide.value('LogService', LogServiceSpy);
+      $provide.value('DataChannelService', DataChannelServiceSpy);
+    })
+
+  });
+
+
+  beforeEach(inject(function (_ActionService_, _$timeout_, _$rootScope_) {
+    ActionService = _ActionService_;
+    $timeout = _$timeout_;
+    $rootScope = _$rootScope_;
+  }));
+
+  afterEach(function () {
+    FeedsServiceSpy.add.calls.reset();
+    FeedsServiceSpy.find.calls.reset();
+    FeedsServiceSpy.allFeeds.calls.reset();
+    FeedSpy.calls.reset();
+    LogEntrySpy.calls.reset();
+    LogServiceSpy.add.calls.reset();
+  });
+
+  it('should define functions', function () {
+    expect(ActionService.enterRoom).toBeFunction();
+    expect(ActionService.leaveRoom).toBeFunction();
+    expect(ActionService.remoteJoin).toBeFunction();
+    expect(ActionService.destroyFeed).toBeFunction();
+    expect(ActionService.ignoreFeed).toBeFunction();
+    expect(ActionService.stopIgnoringFeed).toBeFunction();
+    expect(ActionService.writeChatMessage).toBeFunction();
+    expect(ActionService.publishScreen).toBeFunction();
+    expect(ActionService.toggleChannel).toBeFunction();
+    expect(ActionService.setMedia).toBeFunction();
+  });
+
+
+  // Test enterRoom method
+  describe('enterRoom', function () {
+    var feedId = 1,
+        display = 'd',
+        connection = 'conn';
+
+    it('should call Feed',function () {
+
+      ActionService.enterRoom(feedId, display, connection);
+
+      expect(FeedSpy).toHaveBeenCalledWith({
+        display: display,
+        connection: connection,
+        id: feedId,
+        isPublisher: true
+      });
+    });
+
+    it('should call FeedsService.add',function () {
+      ActionService.enterRoom(feedId, display, connection);
+
+      expect(FeedsServiceSpy.add).toHaveBeenCalledWith(
+        jasmine.objectContaining({fakeFeed: true}),
+        jasmine.objectContaining({main: true})
+      );
+    });
+
+  });
+  // End enterRoom method
+
+  // Test leaveRoom method
+  describe('leaveRoom', function () {
+
+    it('should call FeedsService.allFeeds',function () {
+      spyOn(ActionService, 'destroyFeed');
+      ActionService.leaveRoom();
+
+      expect(FeedsServiceSpy.allFeeds).toHaveBeenCalled();
+    });
+
+    it('should call ActionService.destroyFeed for each returned feed',function () {
+      spyOn(ActionService, 'destroyFeed');
+      ActionService.leaveRoom();
+
+      var feeds = FeedsServiceSpy.allFeeds();
+      expect(ActionService.destroyFeed.calls.count()).toBe(feeds.length);
+    });
+
+
+  });
+  // End leaveRoom method
+
+  // Test publishScreen method
+  describe('publishScreen', function () {
```

Comment:
This case is even worse than above. The 1:1 repetition between implementation and test is taken to the extreme. I will try to come up with a working re-implementation of some of the tests in a more behavior-oriented fashion (also as an exercise to learn jasmine myself).